### DISCUSSION TITLE
Using a same model for both data and API

### DISCUSSION THREAD

**Discussion by g0di** [pydantic/pydantic#7371] [Question] [answered]
Hi folks,

Lastly I wanted to use Pydantic models which could be used for both serializing/deserializing data from my DB (actually Apache parquet files) AND my API (fastapi). I noticed I could use `alias` mechanism in order to somehow map my model attributes to my DB tables columns. 

Something like:
```python
from pydantic import BaseModel, Field

class Region(BaseModel):
    __table__ = "MINT_REGION"

    id: str = Field(alias="REGION_ID")
    name: str = Field(alias="REGION_LABEL")
```
This model allow me to easily parse raw records from my parquet files while also dumping them back again. The alias act as the name of the column in the parquet file. However, when using this model on my fastapi API, the aliases are taken into account as well so the JSONSchema is now using the alias. Also, I can not use a different alias for the JSONSchema and the data model.

I would like to be able to have some kind of two different aliases, one for specifying my column names and another one (eventually) for my API with a an additional function for dumping my model into a raw parquet record like:

```python
from pydantic import BaseModel, Field

class Region(BaseModel):
    __table__ = "MINT_REGION"

    id: str = Field(column="REGION_ID")
    name: str = Field(column="REGION_LABEL", alias="label")

   def to_parquet(self):
       # this is just an example
       return self.model_dump(mode="json", by_alias="column") 
```
I know that there is the `SQLModel` library which extends Pydantic to use `BaseModel` for representing SQL tables and which override the `Field` class. It pretty much looks like to what I want to achieve but for Parquet files. However, the code is apparently not compatible with Pydantic v2 and feels really complicated to put in place for a simple one-short project.

Is there any way to achieve this without having to rely on two different models and do the mapping "by-hand"?

**Comment by sydney-runkle** [comment, marked as answer]
Great question!

We recently added support for this kind of thing with the V2 release. You can read about aliases more [here](https://docs.pydantic.dev/latest/usage/fields/#field-aliases), and you can read more about alias precedence [here](https://docs.pydantic.dev/latest/usage/model_config/#alias-precedence).

Specifically, the `validation_alias` and the `serialization_alias` should come in handy.

**Reply by g0di** [reply]
Thank you for your answer. Unfortunately I've already had a look at the things you mention. It offers a lot of flexibility and possibilities but it does not exactly fit my use case. In particular this is due to the fact that, as soon as you use a model for FastAPI, the aliases are used to generate the openAPI spec. It means I can't use those aliases to adapt the parsing of my data coming from my DB and at the same time use the regular model fields just for my openAPI model.

After having flipped the problem in all direction in my head, I'm not sure I can achieve what I want to do in place with a single model. My best bet is to have two models (one for the API and one the DB) and create a mapping between them. 

The other solution I found recently is also to split my model into a "read" and a "write" models. This way I can use the "read" model for parsing my DB data (validation_alias) and use the field name or seralization_alias for my openAPI model. In the other side I can use the "write" model and a validation_alias for customizing my openAPI schema (when used as a request body) while using the serialization_alias to dump my model to DB. It leads to a bit of duplication but I found it neat and usefull when the read and write model are not exactly identical.

If I would reformulate my original issue I would instead then ask to be able to attach custom user metadata to Pydantic Field objects (as we can do with regular dataclasses) either by subclassing FieldInfo or by simply passing them to Field function. This would give me the possibility to attach my own metadata (column name) to each field and add a custom function for dumping a model to a data model according to those metadata.

**Comment by mikenerone** [comment]
I came here to ask about a similar situation, where I'm also forced to use separate models that differ only in aliasing. I'll chime in with my example, as well. I'm working with a simple RESTful API that uses Pascal-style names for everything. I want aliasing in both directions because I both receive these objects from the API and send them to it, but as per norms, I want my own code to use Pythonic attribute names. However, in some places I construct instances that are going to be _new_ objects to send to the API as creation inputs. That desire to use Pythonic names is just as true in instantiations as it is in attribute accesses.

TBH, I feel like this is what people use aliases for 99% of the time, and I don't really understand why aliasing is designed the way it is:
```python
person = Person.model_validate_json('{"ID": 2, "FirstName": "John", "LastName": "Smith"}')  # JSON I read from the API I don't control, so using aliases for the validation here is fine.
person.last_name -> "Smith"  # In my own code, I get Pythonic names for attribute access. Nice!
person.first_name = "James"  # Ditto for updates.
my_client.update_person(person)  # And aliases works during serialization, too, so the API will understand this. Sweet!
person = my_client.add_person(Person(first_name="John", last_name="Smith"))  # This is my own Python code, so I should get to use my nice Pythonic parameter names here, right? But I can't. These inputs to Person fail validation.
```

So it's that last case specifically where I want the option, in the context of an explicit model instantiation, to use the field names I defined instead of being forced to do the non-Pythonic `Person(FirstName="John", LastName="Smith")`, but as far as I can tell, there's no option for that.

**Reply by g0di** [reply]
Your case can be solved by providing the « populate_by_name » setting to the model configuration

**Reply by mikenerone** [reply]
Well sweet! Somehow I'd missed this, but that's exactly what I needed. Thanks!
